### PR TITLE
COM-314: Refactor theming of `StackBreadcrumbs`

### PR DESCRIPTION
**Pull request by juliawegmayr**
Screenshot 2024-01-22 at 14 19 12

### CODE REVIEW

File: packages/admin/admin/src/stack/breadcrumbs/BreadcrumbsEntry.tsx [hunk 1]
```diff
@@ -1,42 +1,91 @@
...
+            {backButtonUrl && (
+                <>
+                    <IconButton component={BreadcrumbLink} to={backButtonUrl}>
+                        <LevelUp />
+                    </IconButton>
+                    <BackButtonSeparator {...slotProps?.backButtonSeparator} />
+                </>
+            )}
+            {isLastItem ? (
+                <StyledDisabledLink {...slotProps?.styledDisabledLink} variant="body2">
```
Comment (jamesricky) [inline packages/admin/admin/src/stack/breadcrumbs/BreadcrumbsEntry.tsx:81]: The disabled item should not be changed to a link

File: packages/admin/admin/src/stack/breadcrumbs/BreadcrumbsEntry.tsx [hunk 2]
```diff
@@ -1,42 +1,91 @@
...
+import { css, styled, useThemeProps } from "@mui/material/styles";
+import { ThemedComponentBaseProps } from "helpers/ThemedComponentBaseProps";
 import * as React from "react";
 
 import { BreadcrumbItem } from "../Stack";
 import { BreadcrumbLink } from "./BreadcrumbLink";
-import { styles } from "./StackBreadcrumbs.styles";
 
-interface BreadcrumbsEntryProps {
+export type BreadcrumbsEntryClassKey = "backButtonSeparator" | "link" | "disabledLink";
```
Comment (jamesricky) [inline packages/admin/admin/src/stack/breadcrumbs/BreadcrumbsEntry.tsx:11]: I don't think there's a good reason to remove/rename all those class keys... seems linke an unnecessary breaking change to me that now removes the ability to style certain elements entirely.

File: packages/admin/admin/src/stack/breadcrumbs/BreadcrumbsEntry.tsx [hunk 3]
```diff
@@ -1,42 +1,91 @@
...
+    ({ theme }) => css`
+        height: 30px;
+        width: 1px;
+        background-color: ${theme.palette.divider};
+        margin-right: 12px;
+    `,
+);
+
+export interface BreadcrumbsEntryProps
+    extends ThemedComponentBaseProps<{ styledLink: typeof Link; backButtonSeparator: "div"; styledDisabledLink: typeof Link }> {
```
Comment (jamesricky) [inline packages/admin/admin/src/stack/breadcrumbs/BreadcrumbsEntry.tsx:62]: The keys here need to match the values in the `ClassKey` type

**Review by jamesricky — CHANGES_REQUESTED** [review]

File: packages/admin/admin/src/stack/breadcrumbs/BreadcrumbsEntry.tsx [hunk 4]
```diff
@@ -1,42 +1,89 @@
...
-        )}
-    </>
-);
+export function BreadcrumbsEntry(inProps: BreadcrumbsEntryProps) {
+    const { item, isLastItem, backButtonUrl, slotProps } = useThemeProps({ props: inProps, name: "CometAdminBreadcrumbsEntry" });
+    return (
+        <>
+            {backButtonUrl && (
+                <>
+                    <IconButton component={BreadcrumbLink} to={backButtonUrl}>
```
Comment (jamesricky) [inline packages/admin/admin/src/stack/breadcrumbs/BreadcrumbsEntry.tsx:72]: This should be a slot that extends `IconButton`

File: packages/admin/admin/src/stack/breadcrumbs/BreadcrumbsEntry.tsx [hunk 5]
```diff
@@ -1,42 +1,89 @@
...
+    slot: "disabledLink",
+    overridesResolver(_, styles) {
+        return [styles.disabledLink];
+    },
+})(
+    ({ theme }) => css`
+        font-size: 13px;
+        line-height: 14px;
+        font-weight: ${theme.typography.fontWeightMedium};
+        color: ${theme.palette.grey[600]};
```
Comment (jamesricky) [inline packages/admin/admin/src/stack/breadcrumbs/BreadcrumbsEntry.tsx:38]:
```suggestion
```

File: packages/admin/admin/src/stack/breadcrumbs/BreadcrumbsEntry.tsx [hunk 6]
```diff
@@ -1,42 +1,89 @@
...
+        text-decoration-color: currentColor;
+    `,
+);
+
+const StyledDisabledLink = styled(Typography, {
+    name: "CometAdminStackBreadcrumbs",
+    slot: "disabledLink",
+    overridesResolver(_, styles) {
+        return [styles.disabledLink];
```
Comment (jamesricky) [inline packages/admin/admin/src/stack/breadcrumbs/BreadcrumbsEntry.tsx:31]: This should also include the styles of the non-disabled link, as in the previous implementation. 

```suggestion
        return [styles.link, styles.disabledLink];
```

File: packages/admin/admin/src/stack/breadcrumbs/BreadcrumbsEntry.tsx [hunk 7]
```diff
@@ -1,42 +1,89 @@
...
-            </Typography>
-        ) : (
-            <Link to={item.url} component={BreadcrumbLink} className={classes.link} variant="body2">
-                {item.title}
-            </Link>
-        )}
-    </>
-);
+export function BreadcrumbsEntry(inProps: BreadcrumbsEntryProps) {
+    const { item, isLastItem, backButtonUrl, slotProps } = useThemeProps({ props: inProps, name: "CometAdminBreadcrumbsEntry" });
```
Comment (jamesricky) [inline packages/admin/admin/src/stack/breadcrumbs/BreadcrumbsEntry.tsx:67]: `ThemedComponentBaseProps`, its slot-types and `useThemeProps` should only be defined/used in the "primary" component `StackBreadcrumbs` as `BreadcrumbsEntry` is not an exported component but only used internally by `StackBreadcrumbs`.

**Review by jamesricky — CHANGES_REQUESTED** [review]

File: packages/admin/admin/src/stack/breadcrumbs/BreadcrumbsOverflow.tsx [hunk 8]
```diff
@@ -1,32 +1,45 @@
...
-import clsx from "clsx";
+import { css, styled } from "@mui/material/styles";
 import * as React from "react";
 
 import { BreadcrumbItem } from "../Stack";
 import { BreadcrumbLink } from "./BreadcrumbLink";
-import { styles } from "./StackBreadcrumbs.styles";
 
-interface BreadcrumbsOverflowProps {
+export interface BreadcrumbsOverflowProps {
```
Comment (jamesricky) [inline packages/admin/admin/src/stack/breadcrumbs/BreadcrumbsOverflow.tsx:9]: There is no need to export this interface. 

```suggestion
interface BreadcrumbsOverflowProps {
```

**Review by jamesricky — CHANGES_REQUESTED** [review]
Please use the correct type for `slotProps` instead of `any`. 

`StackBreadcrumbsProps["slotProps"]`

**Review by jamesricky — DISMISSED** [review]

File: packages/admin/admin/src/stack/breadcrumbs/BreadcrumbsEntry.tsx [hunk 9]
```diff
@@ -1,42 +1,80 @@
...
+import { IconButton as MuiIconButton, Link, Typography } from "@mui/material";
+import { css, styled } from "@mui/material/styles";
 import * as React from "react";
 
 import { BreadcrumbItem } from "../Stack";
 import { BreadcrumbLink } from "./BreadcrumbLink";
-import { styles } from "./StackBreadcrumbs.styles";
+import { BackButtonSeparator, StackBreadcrumbsProps } from "./StackBreadcrumbs";
+
+const StyledLink = styled(Link, {
```
Comment (johnnyomair) [inline packages/admin/admin/src/stack/breadcrumbs/BreadcrumbsEntry.tsx:10]: I'd prefer if you'd remove the Styled prefixes... They don't really add any meaning to the components. For the name clash on Link you could import the Link as `Link as MuiLink`

**Review by johnnyomair — CHANGES_REQUESTED** [review]

**Review by jamesricky — APPROVED** [review]

**Review by johnnyomair — APPROVED** [review]

**Pull request merged**